### ISSUE TITLE
Can't access some view constructs like gantt, kanban etc.

### ISSUE THREAD

**Issue by oerp-odoo** [odoo/odoo#17645]
**Impacted versions:**
10.0
**Steps to reproduce:**

```javascript
odoo.define('mrp_mo_preliminary.GanttViewMrpPreliminary', function (require) {
"use strict";

var core = require('web.core');

console.log('list' + core.view_registry.get('list'));
console.log('calendar ' + core.view_registry.get('calendar'));
console.log('diagram ' + core.view_registry.get('diagram'));
console.log('form ' + core.view_registry.get('form'));
console.log('graph ' + core.view_registry.get('graph'));
console.log('grid ' + core.view_registry.get('grid'));
console.log('kanban ' + core.view_registry.get('kanban'));
console.log('gantt ' + core.view_registry.get('gantt'));
console.log('tree' + core.view_registry.get('tree'));
console.log('many2many_kanban ' + core.view_registry.get('many2many_kanban'));
console.log('one2many_kanban '+ core.view_registry.get('one2many_kanban'));
console.log('pivot ' + core.view_registry.get('pivot'));

});
```
**Current behavior:**
Output of that log is:

```javascript
list function Class(){if(this.constructor!==OdooClass){throw new Error("You can only instanciate objects with the 'new' operator");}
this._super=null;if(!initializing&&this.init){var ret=this.init.apply(this,arguments);if(ret){return ret;}}
return this;}

web.assets_backend.js:3126 calendar undefined

web.assets_backend.js:3126 diagram undefined

web.assets_backend.js:3126 form function Class(){if(this.constructor!==OdooClass){throw new Error("You can only instanciate objects with the 'new' operator");}
this._super=null;if(!initializing&&this.init){var ret=this.init.apply(this,arguments);if(ret){return ret;}}
return this;}

web.assets_backend.js:3126 graph function Class(){if(this.constructor!==OdooClass){throw new Error("You can only instanciate objects with the 'new' operator");}
this._super=null;if(!initializing&&this.init){var ret=this.init.apply(this,arguments);if(ret){return ret;}}
return this;}

web.assets_backend.js:3126 grid function Class(){if(this.constructor!==OdooClass){throw new Error("You can only instanciate objects with the 'new' operator");}
this._super=null;if(!initializing&&this.init){var ret=this.init.apply(this,arguments);if(ret){return ret;}}
return this;}

web.assets_backend.js:3126 kanban undefined

web.assets_backend.js:3126 gantt undefined

web.assets_backend.js:3126 tree function Class(){if(this.constructor!==OdooClass){throw new Error("You can only instanciate objects with the 'new' operator");}
this._super=null;if(!initializing&&this.init){var ret=this.init.apply(this,arguments);if(ret){return ret;}}
return this;}

web.assets_backend.js:3126 many2many_kanban undefined

web.assets_backend.js:3126 one2many_kanban undefined

web.assets_backend.js:3126 pivot function Class(){if(this.constructor!==OdooClass){throw new Error("You can only instanciate objects with the 'new' operator");}
this._super=null;if(!initializing&&this.init){var ret=this.init.apply(this,arguments);if(ret){return ret;}}
return this;}
```
As you can see some views are undefined even though I see that they were registered there and should be accessible. if I `console.log(core.view_registry.map)` 
It returns this:

```javascript
calendar: function Class()
diagram: function Class()
form: function Class()
gantt: function Class()
graph: function Class()
grid: function Class()
kanban: function Class()
list: function Class()
many2many_kanban: function Class()
one2many_kanban: function Class()
pivot: function Class()
tree: function Class()
```

As you can see every key has its construct. So how come some are undefined?..
My module has dependencies both for `web` and `web_gantt`.

**Expected behavior:**
I expect to access every view that is defined in `web.core.view_registry`.

Is there some gotcha that some views are not accessible directly?

**Update**

It looks it is possible to directly access gantt view without going into `view_registry` using var `GanttView = require('web_gantt.GanttView')`. Though it still does not make sense to register view into registry if it can't be access from there?

**Comment by Yenthe666** [comment]
@ged-odoo could you explain why this happens?

**Comment by ged-odoo** [comment]
@Yenthe666 @oerp-odoo  this is the danger of registries: they do not enforce dependencies.  What happens in your case is that your module only depends on web.core, which is loaded quickly, so your module can be launched quickly.  Gantt view and kanban view (and grid view, and calendar view) are defined in another addon, so they are added at a different time, after your own module is loaded.

So, in short:

1. web.registry is loaded
2. web.core is loaded, it instantiate the view registry
3. all views defined in /web are loaded, and added to the registry
4. your module is loaded, the behavior that you observe happens
5. the views defined in other addons are loaded, and added to the registry
6. you can look at the mapping in the registry, and see that they are indeed in the registry

In conclusion, you should not use the registry to get a reference to view classes, you should directly import the views (var gantt = require('web_gantt.gantt') or something like that).  The registry is meant to be the point of access for the webclient, in runtime, to fetch the proper class.

**Comment by Yenthe666** [comment]
Thank you for the detailed and very good answer.
